### DISCUSSION TITLE
pytest_addoption not working when run from inside a plugin registered inside pytest_configure

### DISCUSSION THREAD

**Discussion by lovetheguitar** [pytest-dev/pytest#9527] [Q&A]
Hello all :) 

I'm writing a pip installable collection of pytest plugins. So I have multiple entry points to different plugin modules. Each module defines different pytest_ methods and fixtures. All working well so far.

Since I'm trying to abstract different hardware, some of which I might have multiple devices of, I was trying to make my plugin more DRY by defining some BasePlugin class and loading it inside pytest_configure. When my BasePlugin class has a pytest_addoption method those entries are not displayed in help neither do they work when passed on the command line.

e.g. in plugin.py
```python
class BasePlugin:
    def pytest_addoption(self, parser):
        """Add command line options."""
        print("This is executed, but does not work")
        parser.addoption("--myinside", help="why does this not work?")


def pytest_configure(config):
    print("pytest_configure gets run")
    config.pluginmanager.register(BasePlugin())


def pytest_addoption(parser):
    """Add command line options outside."""
    print("This is executed, but does not work")
    parser.addoption("--myoutside", help="this works")

```

The flag --myoutside works and is displayed correctly in the help section. 
Whereas, --myinside is not displayed in the help section and when being passed on the cli I get "pytest: error: unrecognized arguments: --myoption ...".

I read this https://docs.pytest.org/en/6.2.x/writing_plugins.html#optionally-using-hooks-from-3rd-party-plugins and https://docs.pytest.org/en/6.2.x/reference.html#pytest.hookspec.pytest_addoption, where it states that "This function should be implemented only in plugins or conftest.py files situated at the tests root directory due to how pytest discovers plugins during startup.". So what I'm trying to do is apparently not supported?

I think I could set all the required API calls in my plugin module's globals directly. Is this the way to go or are there more beautiful ways to solve this? 

Thanks a lot. :)

**Comment by RonnyPfannschmidt** [comment]
pytest_addoption is something that ought to happen waaay before pytest_configure (addoption happens before the config object is even made)

as such there is no way to skip it over

if you have one general base plugin, why not just have that be a always on plugin to begin with

**Reply by lovetheguitar** [reply]
Thanks for the confirmation.
I am not sure what an "always on" plugin is, but I guess mine are?
So in my setup file I specify several pytest11 entry points to different module files, for exmaple one is called hw_multimeter and another hw_loads.

Inside the referenced plugin module files I wanted to support, e.g. different multi meters and loads (all physical hardware, that can be controlled via e.g. ETH). Supporting single instances of e.g. multimeter-keithley-modelxyz and multimeter-agilent-modelabc already works fine. Problem comes with scaling though. I want to support multiple instances (separate physical devices) of e.g. 'multimeter-keithley-modelxyz'. 

So my idea was to have some kind of "plugin factory", to enable using "multimeter-keithley-modelxyz-1" and "multimeter-keithley-modelxyz-2", and so on for multiple hardware devices. Using the a pugin and not a "fixture factory" is imho required to be able to automatically specify available cli, ini flags and fixtures automatically too. 

Example `pytest --help` output:
```bash
  --multimeter-keithley-modelxyz-1-address=MULTIMETER_KEITHLEY_MODELXYZ_1_ADDRESS
                        Specify address for multimeter-keithley-modelxyz-1.
                        Default is AK001A.
  --multimeter-keithley-modelxyz-2-address=MULTIMETER_KEITHLEY_MODELXYZ_2_ADDRESS
                        Specify address for multimeter-keithley-modelxyz-2.
                        Default is AK002A.
  --multimeter-keithley-modelxyz-3-address=MULTIMETER_KEITHLEY_MODELXYZ_3_ADDRESS
                        Specify address for multimeter-keithley-modelxyz-3.
                        Default is AK003A.
  --multimeter-keithley-modelxyz-4-address=MULTIMETER_KEITHLEY_MODELXYZ_4_ADDRESS
                        Specify address for multimeter-keithley-modelxyz-4.
                        Default is AK004A.
  --multimeter-keithley-modelxyz-5-address=MULTIMETER_KEITHLEY_MODELXYZ_5_ADDRESS
                        Specify address for multimeter-keithley-modelxyz-5.
                        Default is AK005A.
```

At the moment I have a BasePlugin class and there I store for example the required cli arguments and later on I inject this into the modules globals() and make sure to inject e.g. pytest_addoption exactly once, which is in turn iterating over the previously injected cli args. Feels quite hacky so far, but seems to work theoretically. Is there anything I missed from the API documentation that could help me achieve my goal? 

Thanks a lot for all the great work you do here everyone! :)

**Reply by RonnyPfannschmidt** [reply]
that sounds awfully lot like instead of a massive mega load of config options on the cli, you want a configfile special to your plugin

**Reply by lovetheguitar** [reply]
Thanks for your input again. Do you have any example plugin/s in mind, so I could have a look? 

I'm a bit lost how I'd structure this best, one file per sub plugin or a general config file plus where/how to translate the config file information into available fixtures.

**Reply by RonnyPfannschmidt** [reply]
without more knowledge on your tes setup its next to impossible to give sound advise

its nto even clear to me how the initial invocations where supposed to work

a few examples  may help
